Adding plugin resource sync. Every time a sync is called, every

albertshau opened the pull request:
resource that is set to go live is found and sent to every
provisioner that has workers assigned to it for that tenant.

albertshau committed:
Adding plugin resource sync. Every time a sync is called, every
resource that is set to go live is found and sent to every
provisioner that has workers assigned to it for that tenant.

albertshau committed:
add missing javadoc and fix checkstyle errors.

albertshau committed:
fix put tenant call to provisioners.

poornachandra commented on server/src/main/java/com/continuuity/loom/http/handler/LoomPluginHandler.java:
```diff
@@ -48,13 +51,17 @@
  */
 @Path("/v1/loom")
 public class LoomPluginHandler extends LoomAuthHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(LoomProvisionerHandler.class);
```
Wrong class in getLogger.

albertshau committed:
explicitly exclude undesired field when serializing resource
collection and sending to provisioner.

albertshau committed:
fix a sync race condition and a bug where the metadata for
resource types that dont get synced are not updated, which can
prevent resources that are unstaged to transition to the
inactive state.

albertshau committed:
make sure stage/unstage on nonexistent resource returns a 404.

poornachandra commented on server/src/main/java/com/continuuity/loom/provisioner/TenantProvisionerService.java:
```diff
@@ -295,7 +303,15 @@ public void writeProvisioner(Provisioner provisioner) throws IOException {
-    lock.acquire();
+    // lock across all tenants to protect against conflicts in setting worker counts for different tenants across
+    // different provisioners
+    tenantLock.acquire();
```
This method takes tenantLock and then syncLock, syncResources takes syncLock and then tenantLock -> this will lead to dead-lock.

albertshau replied:
yeah... I think I'll remove the sync lock entirely and change the way the metadata is synced

poornachandra commented on server/src/main/java/com/continuuity/loom/provisioner/TenantProvisionerService.java:
```diff
@@ -304,17 +320,69 @@ public void rebalanceTenantWorkers(String tenantId) throws IOException, Capacity
+      // to the queue group interface.
+
+      // update tenant provisioners
+      syncProvisionerResources(account.getTenantId(), sync.getResourceCollection());
```
We'll have to make this operation work across Loom Server restarts.

albertshau replied:
agreed.  I think it will be better to do it as a separate pull request because it will require adding functionality to queue groups.

poornachandra commented on server/src/main/java/com/continuuity/loom/provisioner/plugin/ActiveResourceList.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * List of active resources to set on a provisioner. Only used as part of {@link ResourceCollection}.
+ */
+public class ActiveResourceList {
```
Is this only used for JSON conversion?

poornachandra commented on server/src/main/java/com/continuuity/loom/store/provisioner/PluginMetaStoreView.java:
```diff
@@ -1,152 +1,30 @@
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
```
Copyright removed. Also, I noticed that copyright is missing for new files.

albertshau replied:
weird intellij issues, will resolve

poornachandra commented:
Done with review.

albertshau committed:
removing sync lock and explicitly setting which resources are
live instead of trying to coordinate when a stage/unstage has
to happen in relation to a sync.

albertshau committed:
Merge remote-tracking branch 'origin/develop' into feature/resource-sync

Conflicts:
	server/src/main/java/com/continuuity/loom/http/handler/LoomPluginHandler.java
	server/src/main/java/com/continuuity/loom/provisioner/plugin/ResourceService.java
	server/src/test/java/com/continuuity/loom/http/LoomPluginHandlerTest.java
